# Laboratorium 1 : Wprowadzenie do NumPy – podstawowe operacje i porównanie z listami Pythonowymi

## Marcin Ogórkiewicz 120674, grupa 1

### Zadanie 1

In [1]:
import numpy as np
import time as t

python_list = list(range(1000)) #tworzenie listy zawierającej 1000 elementów
pyton_list_sum = sum(python_list) #suma elementów listy
python_list_mul = [x * 2 for x in python_list] #mnożenie każdego elementu listy przez 2
np_array = np.arange(1000) #tworzenie tablicy numpy zawierającej 1000 elementów
np_array_sum = np.sum(np_array) #suma elementów tablicy
np_array_mul = np_array * 2 #mnożenie każdego elementu tablicy przez 2

### Zadanie 2

Czas sumowania elementów listy

In [20]:
list_sum_start = t.perf_counter() #start pomiaru czasu
python_list_sum = sum(python_list) #suma elementów listy
list_sum_stop = t.perf_counter() #koniec pomiaru czasu
list_sum_elapsed = list_sum_stop - list_sum_start #obliczenie czasu potrzebnego do wykonania operacji 
f'{list_sum_elapsed:.8f} s' #sformatowany czas; float, 8 miejsc po przecinku

'0.00032270 s'

Czas mnożenia elementów listy

In [21]:
list_mul_start = t.perf_counter() #start pomiaru czasu
python_list_mul = [x * 2 for x in python_list] #mnożenie każdego elementu listy przez 2
list_mul_stop = t.perf_counter() #koniec pomiaru czasu
list_mul_elapsed = list_mul_stop - list_mul_start #obliczenie czasu potrzebnego do wykonania operacji
f'{list_mul_elapsed:.8f} s' #sformatowany czas; float, 8 miejsc po przecinku

'0.00020890 s'

Czas sumowania elementów tablicy

In [22]:
array_sum_start = t.perf_counter() #start pomiaru czasu
np_array_sum = np.sum(np_array) #suma elementów tablicy
array_sum_stop = t.perf_counter() #koniec pomiaru czasu
array_sum_elapsed = array_sum_stop - array_sum_start #obliczenie czasu potrzebnego do wykonania operacji
f'{array_sum_elapsed:.8f} s' #sformatowany czas; float, 8 miejsc po przecinku

'0.00016080 s'

Czas mnożenia elementów tablicy

In [23]:
array_mul_start = t.perf_counter() #start pomiaru czasu
np_array_mul = np_array * 2 #mnożenie każdego elementu tablicy przez 2
array_mul_stop = t.perf_counter() #koniec pomiaru czasu
array_mul_elapsed = array_mul_stop - array_mul_start #obliczenie czasu potrzebnego do wykonania operacji
f'{array_mul_elapsed:.8f} s' #sformatowany czas; float, 8 miejsc po przecinku

'0.00012330 s'

### Zadanie 3

In [14]:
matrix_2 = np.full((10, 10), 2) #stworzenie macierzy wypełnionej dwójkami
matrix_4 = np.full((10, 10), 4) #stworzenie macierzy wypełnionej czwórkami
suma = np.sum(matrix_2) # Sumowanie macierzy 
suma

200

In [15]:
matrix_2 * matrix_4 # Mnożenie macierzy 

array([[8, 8, 8, 8, 8, 8, 8, 8, 8, 8],
       [8, 8, 8, 8, 8, 8, 8, 8, 8, 8],
       [8, 8, 8, 8, 8, 8, 8, 8, 8, 8],
       [8, 8, 8, 8, 8, 8, 8, 8, 8, 8],
       [8, 8, 8, 8, 8, 8, 8, 8, 8, 8],
       [8, 8, 8, 8, 8, 8, 8, 8, 8, 8],
       [8, 8, 8, 8, 8, 8, 8, 8, 8, 8],
       [8, 8, 8, 8, 8, 8, 8, 8, 8, 8],
       [8, 8, 8, 8, 8, 8, 8, 8, 8, 8],
       [8, 8, 8, 8, 8, 8, 8, 8, 8, 8]])

### Zadanie 4

Czas sumowania elementów listy

In [28]:
big_python_list = list(range(1000000)) # Stworzenie listy 1000000 elementów
list_sum_start = t.perf_counter() #start pomiaru czasu
big_python_list_sum = sum(big_python_list) #suma elementów listy
list_sum_stop = t.perf_counter() #koniec pomiaru czasu
list_sum_elapsed = list_sum_stop - list_sum_start #obliczenie czasu potrzebnego do wykonania operacji 
f'{list_sum_elapsed:.8f} s' #sformatowany czas; float, 8 miejsc po przecinku

'0.06117280 s'

Czas mnożenia elementów listy

In [29]:
list_mul_start = t.perf_counter() #start pomiaru czasu
big_python_list_mul = [x * 2 for x in big_python_list] #mnożenie każdego elementu listy przez 2
list_mul_stop = t.perf_counter() #koniec pomiaru czasu
list_mul_elapsed = list_mul_stop - list_mul_start #obliczenie czasu potrzebnego do wykonania operacji
f'{list_mul_elapsed:.8f} s' #sformatowany czas; float, 8 miejsc po przecinku

'0.26404630 s'

Czas sumowania elementów tablicy

In [30]:
big_np_array = np.arange(1000000) # Stworzenie tablicy z 1000000 elementów
array_sum_start = t.perf_counter() #start pomiaru czasu
big_np_array_sum = np.sum(big_np_array) #suma elementów tablicy
array_sum_stop = t.perf_counter() #koniec pomiaru czasu
array_sum_elapsed = array_sum_stop - array_sum_start #obliczenie czasu potrzebnego do wykonania operacji
f'{array_sum_elapsed:.8f} s' #sformatowany czas; float, 8 miejsc po przecinku

'0.00160340 s'

Czas mnożenia elementów tablicy

In [31]:
array_mul_start = t.perf_counter() #start pomiaru czasu
big_np_array_mul = big_np_array * 2 #mnożenie każdego elementu tablicy przez 2
array_mul_stop = t.perf_counter() #koniec pomiaru czasu
array_mul_elapsed = array_mul_stop - array_mul_start #obliczenie czasu potrzebnego do wykonania operacji
f'{array_mul_elapsed:.8f} s' #sformatowany czas; float, 8 miejsc po przecinku

'0.00246380 s'

Wyniki pomiarów:

Lista 1 000 elementów:
Suma: 0.00032270 s
Mnożenie: 0.00020890 s

Tablica 1 000 elementów:
Suma: 0.00016080 s
Mnożenie: 0.00012330 s

Lista 1 000 000 elementów:
Suma: 0.06117280 s
Mnożenie: 0.26404630 s

Tablica 1 000 000 elementów:
Suma: 0.00160340 s
Mnożenie: 0.00246380 s

Czy różnice w czasie działania są zauważalne przy małych rozmiarach danych?

Różnice w czasie działania nie są zauważalne przy małych rozmiarach danych, aczkolwiek pomiary sugerują, że im bardziej skomplikowana operacja oraz większy zbiór, tym bardziej zwiększa się rozbieżność w czasie potrzebnym na ukończenie operacji. Wtedy numpy prezentuje o wiele większą wydajność.

Jak zmienia się czas działania w miarę zwiększania rozmiaru danych?

Dla list, tysiąckrotnie większy zbiór spowodował, w wypadku dodawania - ponad stukrotne wydłużenie czasu działania, a przy mnożeniu aż tysiąckrotne. Tablice są o wiele wydajniejsze, takie samo zwiększenie liczbności zbioru spowodowało około dziesięciokrotnie dłuższy czas wykonywania.

Jakie inne zalety NumPy można wskazać, oprócz wydajności?

- NumPy pozwala na łatwe wykonywanie matematycznych operacji na całych tablicach, takich jak dodawanie, mnożenie, transponowanie, operacje algebraiczne i macierzowe.

- NumPy jest podstawą dla wielu innych popularnych bibliotek w Pythonie, takich jak Pandas, SciPy, scikit-learn czy TensorFlow, co ułatwia integrację różnych narzędzi w jednym projekcie. Współpraca NumPy z tymi bibliotekami umożliwia tworzenie zaawansowanych aplikacji w dziedzinie analizy danych i uczenia maszynowego.

- NumPy ma prostą składnię.

- NumPy pozwala na tworzenie tablic o różnych typach danych, w tym liczb całkowitych, zmiennoprzecinkowych, logicznych, a także typów zdefiniowanych przez użytkownika.

### Wnioski

Dzięki laboratorium miałem okazję zbadać wydajność biblioteki NumPy. Nie spodziewałem się, że różnice w wydajności między operacjami na tablicach i listach są aż tak duże. Uświadomiło mi to, jak ważne jest korzystanie z tablic przy pracy na wielkich zbiorach danych. Dowiedziałem się również, że biblioteki, z których korzystałem na projekcie z podstaw sztucznej inteligencji bazują na NumPy. Poza tym, miałem okazję przećwiczyć korzystanie z różnych funkcji NumPy.